src/metricscollector/collector/apppoller_test.go

```diff
@@ -0,0 +1,191 @@
+							&events.ContainerMetric{
+								ApplicationId: proto.String("test-app-id"),
+								InstanceIndex: proto.Int32(0),
+								MemoryBytes:   proto.Uint64(1234),
+							},
+						}, nil
+					}
+				})
+
+				It("saves all the metrics to database", func() {
+					Expect(database.SaveMetricCallCount()).To(Equal(3))
+				})
+			})
+
+			Context("when all container metrics are empty", func() {
+				BeforeEach(func() {
+					noaa.ContainerMetricsStub = func(appid string, token string) ([]*events.ContainerMetric, error) {
+						Expect(appid).To(Equal("test-app-id"))
+						Expect(token).To(Equal("bearer test-access-token"))
+
+						return []*events.ContainerMetric{}, nil
+					}
+				})
+
+				It("saves nothings to database", func() {
+					Expect(database.SaveMetricCallCount()).To(BeZero())
+				})
+
+			})
+
+			Context("when container metrics are partially empty", func() {
+				BeforeEach(func() {
+					noaa.ContainerMetricsStub = func(appid string, token string) ([]*events.ContainerMetric, error) {
+						Expect(appid).To(Equal("test-app-id"))
+						Expect(token).To(Equal("bearer test-access-token"))
+
+						if noaa.ContainerMetricsCallCount()%2 == 0 {
+							return []*events.ContainerMetric{}, nil
+						} else {
+							return []*events.ContainerMetric{
+								&events.ContainerMetric{
+									ApplicationId: proto.String("test-app-id"),
+									InstanceIndex: proto.Int32(0),
+									MemoryBytes:   proto.Uint64(1234),
+								},
+							}, nil
+						}
+
+					}
+				})
+
+				It("saves non-empty metrics to database", func() {
+					Expect(database.SaveMetricCallCount()).To(Equal(2))
+				})
+			})
+		})
+
+		Context("when retrieving container metrics all fails", func() {
+
+			BeforeEach(func() {
+				noaa.ContainerMetricsReturns(nil, errors.New("an error"))
+			})
+
+			It("saves nothing to database", func() {
+				Expect(database.SaveMetricCallCount()).To(Equal(0))
+			})
+		})
+
+		Context("when retrieving container metrics partially fails", func() {
+			BeforeEach(func() {
+				noaa.ContainerMetricsStub = func(appid string, token string) ([]*events.ContainerMetric, error) {
+					Expect(appid).To(Equal("test-app-id"))
+					Expect(token).To(Equal("bearer test-access-token"))
+
+					if noaa.ContainerMetricsCallCount()%2 == 0 {
+						return nil, errors.New("an error")
+					} else {
+						return []*events.ContainerMetric{
+							&events.ContainerMetric{
+								ApplicationId: proto.String("test-app-id"),
+								InstanceIndex: proto.Int32(0),
+								MemoryBytes:   proto.Uint64(1234),
+							},
+						}, nil
+					}
+				}
+			})
+
+			It("saves successful results to database", func() {
+				Expect(database.SaveMetricCallCount()).To(Equal(2))
+			})
+
+		})
+
+	})
+
+	Describe("Stop", func() {
+		JustBeforeEach(func() {
+			poller.Stop()
+		})
+
+		Context("when apppoller is started", func() {
+			BeforeEach(func() {
+				poller.Start()
+				time.Sleep(2500 * time.Millisecond)
```

Comment:
if we want to inject, we will need to make the ticker as a argument when creating apppoller.  This looks a little bit wired since it is an internal thing. Did I misunderstand something?

"Otherwise you need to just add a Consistently(noaa.ContainerMetrics) is not invoked"
you mean stop poller just after start?   once the poller started,  ap.pollMetric() will be invoked before getting the stop signal